flake.nix

```diff
@@ -0,0 +1,301 @@
+        # used by Nix by default.
+        buildPhase = ''
+          pnpm build
+        '';
+
+        installPhase = ''
+          rm dist/index-vite.html
+          mv dist/ $out
+        '';
+      });
+
+      b6-js = pkgs.buildNpmPackage {
+        pname = "b6-js";
+        version = "v.0.0.0";
+        src = ./src/diagonal.works/b6/cmd/b6/js;
+        npmDepsHash = "sha256-qzMHjOVRINRZzeTdabz2u+75QrhULS1YzvdeDzWNwLs=";
+
+        buildPhase = ''
+          npm run build
+        '';
+
+        installPhase = ''
+          mkdir $out
+          mv bundle.js $out
+        '';
+      };
+
+      pyproject-file = (pkgs.runCommand "make-pyproject" { } ''
+        substitute ${./python/pyproject.toml.template} $out \
+          --subst-var-by VERSION ''$(${b6-go}/bin/b6-api --pip-version)
+      '');
+
+      pythonProject = pyproject-nix.lib.project.loadPyproject {
+        projectRoot = ./python;
+        pyproject = pkgs.lib.importTOML pyproject-file;
+      };
+
+      get-function-docs = pkgs.writeShellScriptBin "get-function-docs" ''
+        ${b6-go}/bin/b6-api --docs --functions | ${pkgs.lib.getExe pkgs.jq} ".Functions[] | select(.Name == \"''$1\")"
+      '';
+
+      renderedPyProject = pythonProject.renderers.buildPythonPackage {
+        inherit python;
+      };
+
+      b6-py = python.pkgs.buildPythonPackage (renderedPyProject // {
+        # Set the pyproject to be the one we computed via our b6 binary.
+        patchPhase = ''
+          cat ${pyproject-file} > pyproject.toml
+        '';
+
+        nativeBuildInputs = renderedPyProject.nativeBuildInputs ++ [
+          python.pkgs.grpcio-tools
+        ];
+
+        # A couple of hacks necessary to build the proto files and the API.
+        preBuild = ''
+          # Bring in the necessary proto files and the Makefile
+          cp -r ${./proto} ./proto
+          cat ${./Makefile} > some-Makefile
+
+          # Hack: Run the b6-api command outside of the Makefile, using our
+          # version.
+          ${b6-go}/bin/b6-api --functions | python diagonal_b6/generate_api.py > diagonal_b6/api_generated.py
+
+          # Hack: Make the directory structure that the Makefile expects,
+          # then move things to where we want them
+          mkdir python
+          mkdir python/diagonal_b6
+
+          make proto-python -f some-Makefile
+
+          # Cleanup
+          mv python/diagonal_b6/* ./diagonal_b6
+          rm -rf python/diagonal_b6
+        '';
+
+        pythonImportsCheck = [ "diagonal_b6" ];
+      });
+
+
+      pythonEnv = python.withPackages (ps:
+        [
+          b6-py
+
+          # For `make python`
+          ps.grpcio-tools
+
+          # For hacking
+          ps.jupyter
+        ]
+      );
+
+      # Go setup
+      b6-go = with pkgs; gomod2nix.legacyPackages.${system}.buildGoApplication {
+        name = "b6";
+        src = ./src/diagonal.works/b6;
+        buildInputs = [
+          gdal
```

Comment:
Just digging into gdal for a moment, though I guess the point is more general - gdal  changes a lot, and so pinning it to a specific version ends up being important. Right now, that's implicit in the choice of the ubuntu release version we build against (ie mantic, for the moment). Here, if I follow the package links, nix is pinning to a version of gdal that likely isn't the same as the docker file? That doesn't seem good. The same is also true with, eg, the grpc tools (breakages there tend to be less of an issue, though that's beside the point).

Reply:
Pinning is cetrainly a feature of Nix.

There are a few ways to find versions; one is just by browsing the Git history of a particular package - [gdal](https://github.com/NixOS/nixpkgs/commits/nixpkgs-unstable/pkgs/development/libraries/gdal/default.nix?before=5629520edecb69630a3f4d17d3d33fc96c13f6fe+35) - or [via this tool](https://lazamar.co.uk/nix-versions/?channel=nixpkgs-unstable&package=gdal), which kind of works but sometimes seems to return different packages with names "like" gdal (i.e. pythonPackages.gdal or gdalMinimal).

I've added a commit here that fixes the version of gdal to 3.7.1; the version that I see in mantic. Is that right?

Reply:
Yeah, mantic uses gdal 3.7.1 - but as I'm sure you understand - my point here is that if the Dockerfile/Makefile combination continues to be the source of truth, then the nix build will silently break if that's changed (ie to update to a new ubuntu version), and I'd like to either find a way around that. The alternative would be just standardising on nix being the source of truth, but that feels like it's a longer conversation because we'd have to play with it a little. Maybe the path to that is experimenting with the branch you have here.

Reply:
Yes, I suppose if we treat the Docker image here in the repo as the source of (version) truth that will just be duplicated across Nix, because indeed there's no way to reconcile those.

I think in theory one could, as I've made a start on here, just use Nix to build the Docker image too; but I'm neither here nor there on that. I actually think it's also fine to have the Dockerfile because, as you suggest, it might be a bit much to force _anyone_ developing on this to _have_ to use Nix to build anything.

Personally, I'd be happy to just deal with the minor annoyance of matching versions if and when they change; and keeping CI to remind when they might happen.

Just as a random thought; it did occur to me that the gomod2nix updating could actually just be automated away by a CI job anyway; i.e. the CI job could push the changes itself if someone changes go.mod. I'm not sure that's ideal, but it's at least technically feasiable and might solve that problem.

But I agree that there's no escaping the mismatch problem if we keep the Dockerfile _and_ the Nix docker approach.

But just to say, I do consider the Makefile (basically) the source of truth; aside from the minor hacks already mentioned in my other comment (i.e. the flake.nix invokes some targets quite directly there; and could probably use more if we made the binary that the Makefile runs an optional parameter, for example.)